File: android/app/src/googleHome/kotlin/ch/vergasta/plant/GoogleHomeChannel.kt
package ch.vergasta.plant

import android.app.Activity
import androidx.activity.result.ActivityResultCaller
import com.google.home.FactoryRegistry
import com.google.home.Home
import com.google.home.HomeClient
import com.google.home.HomeConfig
import com.google.home.HomeDevice
import com.google.home.matter.common.PermissionsResultStatus
import com.google.home.matter.common.PermissionsState
import com.google.home.matter.standard.HumiditySensorDevice
import com.google.home.matter.standard.RelativeHumidityMeasurement
import com.google.home.matter.standard.TemperatureMeasurement
import com.google.home.matter.standard.TemperatureSensorDevice
import com.google.home.matter.standard.ThermostatDevice
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import kotlin.math.abs
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.launch

/**
 * Les appareils de Google Home, lus par les Home APIs pour la partie Dart
 * (`lib/data/services/google_home_climate_service.dart`).
 *
 * Mêmes trois questions que pour Apple Maison côté iOS, sur un autre canal :
 * ce que la maison laisse faire, la liste des appareils qui mesurent la
 * température ou l'humidité, la mesure de l'un d'eux. Aucune commande
 * envoyée, aucun autre appareil lu.
 *
 * Cette version-ci ne se compile qu'avec le SDK des Home APIs dans le dépôt
 * Maven local (`-PgoogleHome=true`) ; sans lui, c'est la version muette du
 * jeu de sources `noGoogleHome` qui prend sa place. La marche à suivre —
 * projet déclaré, client OAuth, empreinte SHA-1, SDK — est dans
 * `docs/05-technical-architecture.md`, section « Google Home ».
 *
 * Les appels au SDK suivent la documentation des Home APIs ; ils n'ont pas
 * été compilés ici, faute de SDK, et se vérifient à la première
 * construction avec lui.
 */
object GoogleHomeChannel {
  const val NAME = "ch.vergasta.plant/google_home_climate"

  private var client: HomeClient? = null

  /** Une autorisation refusée : un refus n'est pas une question jamais posée. */
  private var refused = false

  private val scope = CoroutineScope(SupervisorJob() + Dispatchers.Main)

  fun register(activity: Activity, messenger: BinaryMessenger) {
    // Ce que l'application lit, et rien de plus : deux grandeurs, trois
    // types d'appareil. Le registre dit au SDK ce qu'il doit savoir rendre.
    val registry =
        FactoryRegistry(
            types = listOf(TemperatureSensorDevice, HumiditySensorDevice, ThermostatDevice),
            traits = listOf(TemperatureMeasurement, RelativeHumidityMeasurement),
        )
    val home = Home.getClient(context = activity, homeConfig = HomeConfig(Dispatchers.IO, registry))
    // La fenêtre d'autorisation rend sa réponse à l'activité : sans un
    // `ActivityResultCaller`, elle ne peut pas être demandée.
    (activity as? ActivityResultCaller)?.let { home.registerActivityResultCallerForPermissions(it) }
    client = home
    MethodChannel(messenger, NAME).setMethodCallHandler { call, result -> handle(call, result) }
  }

  private fun handle(call: MethodCall, result: MethodChannel.Result) {
    when (call.method) {
      "access" -> reply(result, "unavailable") { accessName() }
      "sensors" -> reply(result, emptyList<Map<String, Any>>()) { sensors() }
      "read" -> {
        val id = call.argument<String>("id")
        if (id.isNullOrEmpty()) {
          result.error("bad_args", "sensor id missing", null)
        } else {
          reply(result, null) { read(id) }
        }
      }
      else -> result.notImplemented()
    }
  }

  /**
   * Répond sur le canal, depuis le fil principal. Une maison qui jette vaut
   * la réponse vide : l'écran sait déjà la montrer, et une exception ici
   * emporterait l'application.
   */
  private fun reply(result: MethodChannel.Result, fallback: Any?, body: suspend () -> Any?) {
    scope.launch {
      val value =
          try {
            body()
          } catch (e: Exception) {
            fallback
          }
      result.success(value)
    }
  }

  // MARK: - Autorisation

  private suspend fun state(): PermissionsState? =
      client?.hasPermissions()?.first { it != PermissionsState.PERMISSIONS_STATE_UNINITIALIZED }

  private suspend fun accessName(): String =
      when (state()) {
        PermissionsState.GRANTED -> "authorized"
        PermissionsState.NOT_GRANTED -> if (refused) "denied" else "notDetermined"
        else -> "unavailable"
      }

  /**
   * L'autorisation, demandée une fois si elle n'a pas encore été donnée.
   * Seule la liste des appareils la demande : lire une mesure ne doit pas
   * ouvrir une fenêtre sous les yeux de qui consulte une fiche de plante.
   */
  private suspend fun granted(prompting: Boolean): Boolean {
    val home = client ?: return false
    if (state() == PermissionsState.GRANTED) return true
    if (!prompting) return false
    val answer = home.requestPermissions()
    if (answer.status != PermissionsResultStatus.SUCCESS) {
      refused = answer.status == PermissionsResultStatus.ERROR || answer.status == PermissionsResultStatus.CANCELLED
      return false
    }
    return true
  }

  // MARK: - Appareils

  private suspend fun sensors(): List<Map<String, Any>> {
    val home = client ?: return emptyList()
    if (!granted(prompting = true)) return emptyList()
    val out = mutableListOf<Map<String, Any>>()
    for (structure in home.structures().list()) {
      // La pièce vient de la pièce elle-même : c'est elle qui connaît ses
      // appareils, et le nom qu'elle porte chez la personne.
      val roomNames = mutableMapOf<String, String>()
      for (room in structure.rooms().list()) {
        for (device in room.devices().list()) {
          roomNames[device.id.id] = room.name
        }
      }
      for (device in structure.devices().list()) {
        val (temperature, humidity) = measures(device)
        if (temperature == null && humidity == null) continue
        val item =
            mutableMapOf<String, Any>(
                "id" to device.id.id,
                "name" to device.name,
                "home" to structure.name,
                "temperature" to (temperature != null),
                "humidity" to (humidity != null),
            )
        roomNames[device.id.id]?.takeIf { it.isNotEmpty() }?.let { item["room"] = it }
        out.add(item)
      }
    }
    return out
  }

  private suspend fun device(home: HomeClient, id: String): HomeDevice? {
    for (structure in home.structures().list()) {
      structure.devices().list().firstOrNull { it.id.id == id }?.let {
        return it
      }
    }
    return null
  }

  /**
   * Ce qu'un appareil mesure de l'air d'une pièce : la température et
   * l'humidité relative, sur n'importe lequel de ses types — un thermostat
   * mesure la pièce où il est posé, comme un capteur.
   */
  private suspend fun measures(device: HomeDevice): Pair<Double?, Double?> {
    var temperature: Double? = null
    var humidity: Double? = null
    for (type in device.types().list()) {
      for (trait in type.traits()) {
        when (trait) {
          is TemperatureMeasurement -> temperature = temperature ?: trait.measuredValue?.toDouble()
          is RelativeHumidityMeasurement -> humidity = humidity ?: trait.measuredValue?.toDouble()
          else -> {}
        }
      }
    }
    return temperature to humidity
  }

  // MARK: - Mesure

  private suspend fun read(id: String): Map<String, Any> {
    val out = mutableMapOf<String, Any>("at" to System.currentTimeMillis())
    val home = client
    if (home == null || !granted(prompting = false)) {
      out["error"] = "unauthorized"
      return out
    }
    val device = device(home, id)
    if (device == null) {
      out["error"] = "unknown device"
      return out
    }
    val (temperature, humidity) = measures(device)
    temperature?.let { out["temperature"] = celsius(it) }
    humidity?.let { out["humidity"] = percent(it) }
    if (temperature == null && humidity == null) out["error"] = "no measurement"
    return out
  }

  /**
   * Matter compte en centièmes de degré ; selon sa version, le SDK rend déjà
   * des degrés. Une pièce ne dépasse pas cent degrés : au-delà, la valeur
   * est en centièmes, et se divise par cent.
   */
  private fun celsius(raw: Double): Double = if (abs(raw) > 100) raw / 100 else raw

  /** Même règle pour l'humidité relative, comptée en centièmes de pour cent. */
  private fun percent(raw: Double): Double = if (raw > 100) raw / 100 else raw
}

File: ios/Runner/GoogleHomeChannel.swift
import Flutter

#if canImport(GoogleHomeSDK)
  import GoogleHomeSDK
  import GoogleHomeTypes
#endif

/// Les appareils de Google Home, lus par les Home APIs pour la partie Dart
/// (`lib/data/services/google_home_climate_service.dart`).
///
/// Mêmes trois questions que pour Apple Maison, sur un autre canal : ce que
/// la maison laisse faire, la liste des appareils qui mesurent la
/// température ou l'humidité, la mesure de l'un d'eux. Aucune commande
/// envoyée, aucun autre appareil lu.
///
/// À la différence de HomeKit, les Home APIs ne sont pas dans le système :
/// leur SDK se télécharge depuis la console Google Home et s'ajoute au
/// projet. Sans lui, `canImport` est faux, le canal ne s'enregistre pas, et
/// Dart n'obtient rien — `AppConfig.googleHomeEnabled` garde alors la maison
/// hors de l'écran. Le reste de la marche à suivre — projet, client OAuth,
/// App Attest, groupe d'applications — est dans
/// `docs/05-technical-architecture.md`, section « Google Home ».
///
/// Les appels au SDK suivent la documentation des Home APIs ; faute de SDK
/// ici, ils n'ont pas été compilés, et se vérifient à la première
/// construction avec lui.
final class GoogleHomeChannel {
  static let name = "ch.vergasta.plant/google_home_climate"

  /// Gardé en vie tant que l'application tourne : la session Google Home
  /// s'ouvre une fois, et le canal ne retient pas son gestionnaire.
  private static var shared: GoogleHomeChannel?

  #if canImport(GoogleHomeSDK)
    /// La session ouverte, et le consentement refusé — un refus n'est pas
    /// une question jamais posée, et l'écran ne dit pas la même chose.
    fileprivate static var session: Home?
    fileprivate static var refused = false
    fileprivate static var configured = false
  #endif

  static func register(with messenger: FlutterBinaryMessenger) {
    #if canImport(GoogleHomeSDK)
      let instance = GoogleHomeChannel()
      shared = instance
      let channel = FlutterMethodChannel(name: name, binaryMessenger: messenger)
      channel.setMethodCallHandler { call, result in instance.handle(call, result: result) }
    #else
      // Sans le SDK, rien à enregistrer : le canal reste absent, et Dart lit
      // une maison vide au lieu d'attendre une réponse qui ne viendrait pas.
      _ = messenger
    #endif
  }
}

#if canImport(GoogleHomeSDK)

  extension GoogleHomeChannel {
    /// Le projet déclaré dans la console Google Home, lu dans `Info.plist` :
    /// rien de tout cela n'est un secret, mais rien de tout cela n'a sa
    /// place en dur dans le code.
    fileprivate struct Project {
      let clientID: String
      let teamID: String
      let appGroup: String?

      init?() {
        let info = Bundle.main.infoDictionary ?? [:]
        guard let clientID = info["GoogleHomeClientID"] as? String, !clientID.isEmpty,
          let teamID = info["GoogleHomeTeamID"] as? String, !teamID.isEmpty
        else { return nil }
        self.clientID = clientID
        self.teamID = teamID
        self.appGroup = info["GoogleHomeAppGroup"] as? String
      }
    }

    fileprivate func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
      switch call.method {
      case "access":
        Task { result(await self.accessName()) }
      case "sensors":
        Task { result(await self.sensors()) }
      case "read":
        guard let args = call.arguments as? [String: Any], let id = args["id"] as? String, !id.isEmpty else {
          result(FlutterError(code: "bad_args", message: "sensor id missing", details: nil))
          return
        }
        Task { result(await self.read(id: id)) }
      default:
        result(FlutterMethodNotImplemented)
      }
    }

    // MARK: - Session

    /// La maison, une fois ouverte.
    ///
    /// `restoreSession` ne demande rien ; c'est `connect` qui ouvre la
    /// fenêtre de consentement, et seule la liste des appareils la
    /// déclenche — lire une mesure ne doit pas la poser sous les yeux de qui
    /// consulte une fiche de plante.
    private func home(prompting: Bool) async -> Home? {
      if let open = Self.session { return open }
      guard let project = Project() else { return nil }
      if !Self.configured {
        Home.configure {
          $0.clientID = project.clientID
          $0.teamID = project.teamID
          if let group = project.appGroup { $0.sharedAppGroup = group }
        }
        Self.configured = true
      }
      if let restored = await Home.restoreSession() {
        Self.session = restored
        return restored
      }
      guard prompting else { return nil }
      do {
        let opened = try await Home.connect()
        Self.session = opened
        return opened
      } catch {
        Self.refused = true
        return nil
      }
    }

    private func accessName() async -> String {
      guard Project() != nil else { return "unavailable" }
      if await home(prompting: false) != nil { return "authorized" }
      return Self.refused ? "denied" : "notDetermined"
    }

    // MARK: - Appareils

    /// Les appareils qui mesurent la température ou l'humidité de l'air,
    /// avec leur pièce et leur maison.
    private func sensors() async -> [[String: Any]] {
      guard let home = await home(prompting: true) else { return [] }
      var out: [[String: Any]] = []
      do {
        for structure in try await home.structures().list() {
          // La pièce vient de la pièce elle-même : c'est elle qui connaît
          // ses appareils, et le nom qu'elle porte chez la personne.
          var roomNames: [String: String] = [:]
          for room in try await structure.rooms().list() {
            for device in try await room.devices().list() {
              roomNames[device.id.id] = room.name
            }
          }
          for device in try await structure.devices().list() {
            let (temperature, humidity) = await measures(of: device)
            if temperature == nil && humidity == nil { continue }
            var item: [String: Any] = [
              "id": device.id.id,
              "name": device.name,
              "home": structure.name,
              "temperature": temperature != nil,
              "humidity": humidity != nil,
            ]
            if let room = roomNames[device.id.id], !room.isEmpty { item["room"] = room }
            out.append(item)
          }
        }
      } catch {
        return out
      }
      return out
    }

    private func device(id: String, in home: Home) async -> HomeDevice? {
      guard let structures = try? await home.structures().list() else { return nil }
      for structure in structures {
        guard let devices = try? await structure.devices().list() else { continue }
        if let found = devices.first(where: { $0.id.id == id }) { return found }
      }
      return nil
    }

    /// Ce qu'un appareil mesure de l'air d'une pièce, par les types qui le
    /// portent : le capteur de température, l'hygromètre, et le thermostat,
    /// qui mesure la pièce où il est posé. Un type de plus s'ajoute ici.
    private func measures(of device: HomeDevice) async -> (temperature: Double?, humidity: Double?) {
      var temperature: Double?
      var humidity: Double?
      if let sensor = await device.types.get(Matter.TemperatureSensorDeviceType.self) {
        temperature = number(sensor.matterTraits.temperatureMeasurementTrait?.attributes.measuredValue)
      }
      if let sensor = await device.types.get(Matter.HumiditySensorDeviceType.self) {
        humidity = number(sensor.matterTraits.relativeHumidityMeasurementTrait?.attributes.measuredValue)
      }
      if let thermostat = await device.types.get(Matter.ThermostatDeviceType.self) {
        temperature = temperature ?? number(thermostat.matterTraits.temperatureMeasurementTrait?.attributes.measuredValue)
        humidity = humidity ?? number(thermostat.matterTraits.relativeHumidityMeasurementTrait?.attributes.measuredValue)
      }
      return (temperature, humidity)
    }

    private func number(_ value: Any?) -> Double? {
      if let n = value as? NSNumber { return n.doubleValue }
      if let d = value as? Double { return d }
      if let i = value as? Int { return Double(i) }
      if let f = value as? Float { return Double(f) }
      return nil
    }

    // MARK: - Mesure

    private func read(id: String) async -> [String: Any] {
      var out: [String: Any] = ["at": Int(Date().timeIntervalSince1970 * 1000)]
      guard let home = await home(prompting: false) else {
        out["error"] = "unauthorized"
        return out
      }
      guard let device = await device(id: id, in: home) else {
        out["error"] = "unknown device"
        return out
      }
      let (temperature, humidity) = await measures(of: device)
      if let t = temperature { out["temperature"] = Self.celsius(t) }
      if let h = humidity { out["humidity"] = Self.percent(h) }
      if temperature == nil && humidity == nil { out["error"] = "no measurement" }
      return out
    }

    /// Matter compte en centièmes de degré ; selon sa version, le SDK rend
    /// déjà des degrés. Une pièce ne dépasse pas cent degrés : au-delà, la
    /// valeur est en centièmes, et se divise par cent.
    fileprivate static func celsius(_ raw: Double) -> Double { abs(raw) > 100 ? raw / 100 : raw }

    /// Même règle pour l'humidité relative, comptée en centièmes de pour cent.
    fileprivate static func percent(_ raw: Double) -> Double { raw > 100 ? raw / 100 : raw }
  }

#endif
